### DISCUSSION TITLE
Docker support

### DISCUSSION THREAD

**Discussion by GGLabCenter** [waydroid/waydroid#865] [Q&A]
Hi,
I wanna run waydroid into a docker container. Is this possible? (I did not find any mention on the website)

Thanks in advance!

**Comment by Tiritibambix** [comment]
I'd love to too. Any further information on potential docker support ?
Thanks

**Reply by MuajAmin** [reply]
Just use redroid

**Comment by TriMoon** [comment]
Maybe see: https://stackoverflow.com/questions/43741660/is-there-a-way-how-to-run-lxd-inside-docker
- https://github.com/micw/docker-lxc-demo
- https://github.com/micw/docker-lxc

**Comment by MobCode100** [comment]
I've spend most of my time for 3 days trying to setup Waydroid in a Docker container, well unfortunately, stuck on this error #967. @TriMoon mentioned a link showing that it's possible to run LXC in Docker but it requires the time to study both Waydroid and docker-lxc source code to make necessary changes which I don't have motivation for that (as running both lxc and docker side by side is just fine). Plus, requiring Docker container to run in privileged mode and some distros need to mount cgroup which both are security concerns.

So currently, it's just not possible to get Waydroid running by simply installing it inside a Docker container. If running in Docker container is really necessary, anyone can create an issue and hopefully developers will make it someday.

**Reply by regulad** [reply]
Hey @MobCode100, would you consider sharing the progress you made? I have a lot of experience massaging Docker and docker and would like to try to take this challenge on.

**Reply by MobCode100** [reply]
My memory is quite hazy for this one, I don't think I could help you that much, sorry for that. Anyway, feel free to ask me anything. Most of what I did was directly install Waydroid inside docker container and troubleshoot every error I encounter but stuck at that one error I mentioned above. I lack the experience using LXC, maybe there's some clues if I tap into LXC more. I might try explore it for fun one day, when I'm free. Good luck on your endeavor!

**Comment by Geremia** [comment]
I'd like to see this, too. What's the status of a WayDroid docker container?

**Comment by regulad** [comment]
No worries, I ended up finding another project
https://github.com/remote-android/redroid-doc that is built around running
Android in docker. I think it will work quite well for my application.

**Reply by MobCode100** [reply]
Thank you for sharing! never heard of this one. Might try this one day.

**Comment by realyukii** [comment]
If waydroid is a container, Why put a container inside a container?

**Reply by JavierAgueroCL** [reply]
To prevent junk on uninstall

**Comment by supechicken** [comment]
Hello there, I have just managed to boot LineageOS 23.0 on Docker (not Redroid, built from Waydroid source)

See https://github.com/WayDroid-ATV/waydroid-docker-builds for detailed instructions

<img width="2880" height="1800" alt="2026-05-13-234404_hyprshot" src="https://github.com/user-attachments/assets/1eaa6c2d-37ca-4c55-beef-0f14a854f610" />

**Reply by kuolemaaa** [reply]
closed source?

**Reply by supechicken** [reply]
I don't do closed source stuff, at least for now

**Reply by kuolemaaa** [reply]
can you provide the Dockerfile please?

**Reply by israelvf** [reply]
Hi, thanks for sharing this.

I have been looking into this image as well, because I was also curious about how the Docker build was done.

@kuolemaaa, from what I could gather by inspecting `supechicken/waydroid:lineage-23.2`, it does not look like there is a complex Dockerfile involved, at least not in the final image. The image history shows a single layer with `Imported from -`, and `docker inspect` shows the entrypoint as:

```json
[
  "/init",
  "ro.hardware.hwcomposer=redroid",
  "ro.adb.secure=0"
]
```

So my current understanding is that the Docker image appears to be an imported Android rootfs, with Android's `/init` running directly as PID 1, rather than a regular Linux base image that installs and starts Waydroid through a traditional Dockerfile.

The rootfs itself contains the expected Android/LineageOS layout (`system/`, `vendor/`, `apex/`, `odm/`, etc.), plus Waydroid-specific components such as `waydroid-preinit`, `waydroid-init`, `init.waydroid.rc`, `vendor/waydroid.prop`, and the Waydroid HALs. It also includes `hwcomposer.redroid.so`, which seems to explain the use of `ro.hardware.hwcomposer=redroid` for the headless/scrcpy setup.

So maybe the most useful thing to document is not necessarily a Dockerfile, but the process used to generate/assemble the Android rootfs before importing it into Docker. My guess is that the flow is something like:

```text
Waydroid/LineageOS source
→ WayDroid-ATV patches/device/vendor trees
→ generated system/vendor images
→ extracted/assembled Android rootfs
→ redroid hwcomposer/headless pieces
→ docker import with /init as entrypoint
```

@supechicken, does that match how you built it?
